### PR TITLE
Cassette-type functionality with Mixtape

### PR DESCRIPTION
**Pull request by tshort**
It'd be nice to be able to change functionality when statically compiling. This PR includes code from Mixtape to do that.

In the bigger picture, there are different options to support this feature:
1. Wait for the AbstractInterpreter and the compiler-plugin interface to settle. I haven't seen updates to https://github.com/JuliaLang/julia/pull/41632 in a while.
2. Add functionality like this to GPUCompiler.
3. [This PR] Add functionality here, and switch it later based on advancements in option 1 or 2.

Pinging @femtomc for opinions...

Notes:
* Mixtape itself doesn't work on Julia v1.7, so I pulled in and adapted it's code.
* This doesn't include the `optimize!` (post-inference) part of Mixtape.
* This currently only works for `generate_obj`.

There are lot's of possibilities to use this with StaticTools. This PR includes an example where Strings are replaced by StaticStrings.

### CODE REVIEW

**Comment by femtomc** [comment]
@tshort I really like this idea - this was something I tried early on, but I didn't understand the static compilation process very well.

Re -- @vchuravy: is it possible that we can chat about fixing the OrcV2 linking questions in the next few weeks?

**Comment by femtomc** [comment]
Re -- @tshort: if `StaticCompiler` is using GPUCompiler anyways (and doesn't intercept the native execution engine somehow), then the plugin interface won't necessarily offer anything over Mixtape.

**Comment by femtomc** [comment]
The biggest issue with Mixtape is that it uses AbstractInterpreter -- which has the problem that composition is not easy. So e.g. running Mixtape passes and then doing an AD pass is something which there's not really an idiom for. I'd really like to alleviate this issue in general (but I've been busy doing other research).

**Comment by MasonProtter** [comment]
Nice, this is a great idea.

**Comment by tshort** [comment]
The main holdup this Mixtape support is getting it to work with Julia v1.8. I haven't gotten far on that.

**Comment by tshort** [comment]
@MasonProtter, with [this](https://github.com/JuliaCompilerPlugins/Mixtape.jl/blob/681c48b67e8877e08983810a6c0dc946b713fa99/src/Mixtape.jl#L154) one tweak from your Orc v2 PR, the StaticCompiler tests progress much better. Only nightly fails immediately.

**Comment by MasonProtter** [comment]
Okay, that’s promising!

**Comment by brenhinkeller** [comment]
I wonder if this coul work on Nightly now after the update to GPUCompiler 0.16?

**Comment by tshort** [comment]
> I wonder if this could work on Nightly now after the update to GPUCompiler 0.16? Or is GPUCompiler working merely a necessary but not sufficient condition for Mixtape.jl working?

I'm hopeful, too. I gave it a quick try, but it needed more work than I had time for.

**Comment by tshort** [comment]
Here are some updates on Mixtape support. This branch is caught up to master, and tests pass. I used the AbstractInterpreter straight out of GPUCompiler with minor tweaks. That should make it easier to keep up with GPUCompiler, and if it works, we may want to upstream it at some point. 

Some things that still need work:
* The caching of code needs work. Right now, code is getting cached, so if you change functionality of the mixtape conversion or you change your Mix, it'll grab the old code from the cache. Clearing the cache is a dirty (but easy) way around this.
* It could use more testing.
* For some reason I've yet to figure out, the example in `examples/mixtape-statictools/mixtape-fun.jl` fails when manipulating the CodeInfo ([here](https://github.com/tshort/StaticCompiler.jl/blob/a878b500e8e4b9d10a704c593c070ac14bb37419/examples/mixtape-statictools/mixtape-fun.jl#L44)).

**Comment by tshort** [comment]
I think the Mixtape support is ready to merge. But, note that it is experimental and hasn't been tested much. Any other comments prior to merging? @brenhinkeller or @MasonProtter?

**Comment by brenhinkeller** [comment]
Looks like tests are all passing!

**Comment by MasonProtter** [comment]
Great that you got this working. I think most of the functionality here is already covered by the new `@device_override` stuff, but I think this is probably useful to have if anyone needs to do more complex IR transformations than just method replacement.

File: test/fib.o
Comment (MasonProtter) [inline test/fib.o:1]: I don't think this should be committed to the repo, it should be `git rm`'d.

File: test/fib.so
Comment (MasonProtter) [inline test/fib.so:1]: I don't think this should be committed to the repo, it should be `git rm`'d.

File: test/libfoo.o
Comment (MasonProtter) [inline test/libfoo.o:1]: I don't think this should be committed to the repo, it should be `git rm`'d.

File: test/libfoo.so
Comment (MasonProtter) [inline test/libfoo.so:1]: I don't think this should be committed to the repo, it should be `git rm`'d.

**Review by MasonProtter — CHANGES_REQUESTED** [review]

**Comment by tshort** [comment]
> Great that you got this working. I think most of the functionality here is already covered by the new `@device_override` stuff, but I think this is probably useful to have if anyone needs to do more complex IR transformations than just method replacement.

Yeah, overriding methods is sure easier.

File: src/StaticCompiler.jl [hunk 1]
```diff
@@ -136,18 +140,27 @@ shell> tree \$path
...
     tm = GPUCompiler.llvm_machine(external ? ExternalNativeCompilerTarget() : NativeCompilerTarget())
-    job, kwargs = native_job(f, tt, external; name, kwargs...)
     #Get LLVM to generated a module of code for us. We don't want GPUCompiler's optimization passes.
+    job = CompilerJob(NativeCompilerTarget(), 
+                      FunctionSpec(f, tt, false, name), 
+                      StaticCompilerParams(; 
+                                            opt = true, 
+                                            mixtape = mixtape,
+                                            optlevel = Base.JLOptions().opt_level))
+    GPUCompiler.eval(:(current_job = $job))
```
[the remark this answers is not in the document]
Comment (MasonProtter) [reply src/StaticCompiler.jl:158]: I just ran the testsuite locally with this line removed and everything passed, so I think we can remove this.

**Review by MasonProtter — APPROVED** [review]
Looks good!

**Pull request merged**